packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx

```diff
@@ -0,0 +1,81 @@
+import React, { ComponentPropsWithoutRef, CSSProperties, ElementType } from 'react';
+import clsx from 'clsx';
+import useSwitch, { UseSwitchProps } from './useSwitch';
+import classes from './switchUnstyledClasses';
+import ButtonUnstyled from '../ButtonUnstyled/ButtonUnstyled';
+
+export interface SwitchUnstyledProps<
+  TRoot extends ElementType,
+  TThumb extends ElementType,
+  TInput extends ElementType,
+> extends UseSwitchProps {
+  components?: {
+    Root?: TRoot;
+    Thumb?: TThumb;
+    Input?: TInput;
+  };
+  componentsProps?: {
+    root?: ComponentPropsWithoutRef<TRoot>;
+    thumb?: ComponentPropsWithoutRef<TThumb>;
+    input?: ComponentPropsWithoutRef<TInput>;
+  };
+}
+
+const minimalRootStyles: CSSProperties = {
+  position: 'relative',
+};
+
+const minimalInputStyles: CSSProperties = {
+  position: 'absolute',
+  width: '100%',
+  height: '100%',
+  top: 0,
+  left: 0,
+  opacity: 0,
+  zIndex: 1,
+  margin: 0,
+};
+
+const SwitchUnstyled = function SwitchUnstyled<
+  TRoot extends ElementType = 'span',
```

Comment:
Besides having typed `componentsProps`, I wanted to use the type system to enforce constraints on the inner components. In case of SwitchUnstyled, this means for example having `TRoot` with `children` prop. However, setting `TRoot extends ElementType<{children?: ReactNode}>` doesn't do the trick. I'll investigate this deeper and would appreciate anyone's (@eps1lon ?) ideas.

Reply:
> However, setting TRoot extends ElementType<{children?: ReactNode}> doesn't do the trick. I'll investigate this deeper and would appreciate anyone's (@eps1lon ?) ideas.

When you say it doesn't work do you mean that I could pass an element type that doesn't implement `children`? What happens if you constrain it further to `TRoot extends ElementType<children: JSX.Element>`? 

Or try `React.JSXElementConstructor` instead

Reply:
With `TRoot extends ElementType<{ children?: ReactNode }>` TypeScript gives up where the TRoot is instantiated saying "Expression produces a union type that is too complex to represent". I'm going to spend some time on this today so hopefully I'll find a solution.

Reply:
Yeah, just be careful not to spend too much time on it. These issues are very likely to turn out as not supported in TypeScript.